draft-ietf-tls-esni.md

```diff
@@ -255,97 +281,92 @@ extension, defined as follows:
+ClientEncryptedCH structure:
 
 ~~~~
    struct {
        CipherSuite suite;
-       KeyShareEntry key_share;
        opaque record_digest<0..2^16-1>;
+       opaque enc<0..2^16-1>;
        opaque encrypted_ch<0..2^16-1>;
    } ClientEncryptedCH;
 ~~~~
 
 suite
 : The cipher suite used to encrypt ClientHelloInner.
 
-key_share
-: The KeyShareEntry carrying the client's public ephemeral key share
-used to derive the ESNI key.
-
 record_digest
-: A cryptographic hash of the ESNIConfig structure from which the ESNI
+: A cryptographic hash of the ECHOConfig structure from which the ECHO
 key was obtained, i.e., from the first byte of "version" to the end
 of the structure.  This hash is computed using the hash function
 associated with `suite`.
 
-encrypted_sni
-: The serialized ClientHelloInner structure, AEAD-encrypted using
+enc
+: The HPKE encapsulated key, used by servers to decrypt the corresponding
+encrypted_ch field.
+
+encrypted_ch
+: The serialized and encrypted ClientHelloInner structure, AEAD-encrypted using
 cipher suite "suite" and the key generated as described below.
 {:br}
 
-
-If the server accepts ESNI, then does not send this extension.
-If it rejects ESNI, then it sends the following structure in
+If the server accepts ECHO, it does not send this extension.
+If it rejects ECHO, then it sends the following structure in
 EncryptedExtensions:
 
 ~~~
    struct {
-       ESNIConfig retry_keys<1..2^16-1>;
+       ECHOConfig retry_keys<1..2^16-1>;
    } ServerEncryptedCH;
 ~~~
 
 retry_keys
-: One or more ESNIConfig structures containing the keys that the client should use on
-subsequent connections to encrypt the ClientESNIInner structure.
+: One or more ECHOConfig structures containing the keys that the client should use on
+subsequent connections to encrypt the ClientECHOInner structure.
 
-This protocol also defines the "esni_required" alert, which is sent by the
+This protocol also defines the "echo_required" alert, which is sent by the
 client when it offered an "encrypted_server_name" extension which was not
 accepted by the server.
 
+# The "echo_nonce" extension {#echo-nonce}
 
-# The "esni_nonce" extension {#esni-nonce}
-
-When using ESNI, the client MUST also add an extension of type
-"esni_nonce" to the ClientHelloInner (but not to the outer
-ClientHello).
-
-The encrypted SNI is carried in an "encrypted_server_name"
-extension, defined as follows:
+When using ECHO, the client MUST also add an extension of type
+"echo_nonce" to the ClientHelloInner (but not to the outer
+ClientHello). This extension is defined as follows:
 
 ~~~
    enum {
-       esni_nonce(0xffce), (65535)
+       echo_nonce(0xffce), (65535)
    } ExtensionType;
 
    struct {
        uint8 nonce[16];
-   } ESNINonce;
+   } ECHONonce;
 ~~~~
 
 nonce
-: A random 16-octet value generated by the client and echoed by the
-server.
+: A 16-byte nonce exported from the HPKE encryption context. See {{send-echo}}
+for details about its computation.
```

Comment:
Why is the client adding this, given that the peer can compute it? Is that to inject it into the transcript?

Reply:
Yeah, it replaces the esni_nonce value we put in CHInner.

Reply:
We *could* rely on different CH.random values across inner/outer CHs. That would have the same transcript flip/flop effect. I prefer this as it binds the HPKE operation to the transcript pretty strongly.